### PR TITLE
Social Links Block: Add icon and background color options

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
## Description

Added options to the social links block allowing a user to specify the icon or background colors and have that apply to all the social links contained within it.

_Note: Saving the styles in the social link blocks will be done if this general approach is continued instead of switching to the useColors hook if possible._

https://github.com/WordPress/gutenberg/issues/21605

## How has this been tested?

Only manual testing.
WordPress: 5.5.1

#### Testing Instructions
1. Create a post adding social links block to it and adding urls for multiple platforms before saving.
2. Apply this PR
3. Reload the page, block should still be valid
4. Select the social links block
5. Select icon and background colors from the Inspector controls (include custom colors in testing)
6. Save post and reload editor. Colors should be maintained.
7. Switch style to logos only. The background color options should not be shown and when switching styles back the background color will have been cleared.

## Screenshots 
SocialIcons-Colors

## Types of changes

New feature. Enhancement.

## Next Steps
- [x] Add class names and styles to social link saving for front end display
- [x] ~~Switch to using `__experimentalUseColors` if its possible to achieve same editor functionality. (didn't work for me)~~
- [ ] Refine approach

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
@apeatling and @glendaviesnz these color options are working for me although I feel there should be a better way of achieving it than I've come up with. I'd appreciate any suggestions or feedback you have.

File: packages/block-library/src/social-links/edit.js [hunk 1]
```diff
@@ -22,17 +27,54 @@ const TEMPLATE = [
...
+export function SocialLinksEdit( props ) {
+	const {
+		attributes: { className },
+		backgroundColor,
+		setBackgroundColor,
+		iconColor,
+		setIconColor,
+	} = props;
+
+	const logosOnly = className.indexOf( 'is-style-logos-only' ) >= 0;
```
Comment (apeatling) [inline packages/block-library/src/social-links/edit.js:39]: When adding the block first time, the className is undefined so this `indexOf` call causes an error. I think you'll need an `className &&` prepended to check it exists.

File: packages/block-library/src/social-link/edit.js [hunk 2]
```diff
@@ -27,13 +29,35 @@ import { keyboardReturn } from '@wordpress/icons';
  */
 import { getIconBySite, getNameBySite } from './social-list';
 
-const SocialLinkEdit = ( { attributes, setAttributes, isSelected } ) => {
+const SocialLinkEdit = ( {
+	attributes,
+	setAttributes,
+	isSelected,
+	context,
+	rgbIconColor,
```
Comment (apeatling) [inline packages/block-library/src/social-link/edit.js:37]: Could these just be `iconColor` and `backgroundColor` or does the `usesContext` mean they have to be named differently?

Comment (apeatling) [reply packages/block-library/src/social-link/edit.js:37]: Oh I see the naming clash now in edit.

**Review by apeatling — CHANGES_REQUESTED** [review]

**Comment by apeatling** [comment]
This functions well. I think there should be a `contrast-checker` component in place for the checking the background and icon color combination selected.

File: packages/block-library/src/social-links/edit.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/social-links/edit.js]: Is optional chaining (?.) ok to use here? It works and it is a little more concise. Happy to change it to use `className &&` if that is preferred.

**Comment by aaronrobertshaw** [comment]
I've added a contrast checker although it doesn't really handle the default backgrounds that vary in color by the social media platform. 

This PR has also been rebased and conflicts resolved after the addition of the "open in new tab" feature for social links. I also had to change the approach a little as for the moment colors cannot be retrieved from the general settings ( https://github.com/WordPress/gutenberg/pull/25419 ). There is a PR to re-add the default colors and gradients ( https://github.com/WordPress/gutenberg/pull/25523 ) however updating this to leverage the `__experimentalUseEditorFeature` approach might be better long term.

File: packages/block-library/src/social-link/index.php [hunk 3]
```diff
@@ -16,11 +16,10 @@
  */
 function render_block_core_social_link( $attributes, $content, $block ) {
 	$open_in_new_tab = $block->context['openInNewTab'];
-
-	$service    = ( isset( $attributes['service'] ) ) ? $attributes['service'] : 'Icon';
-	$url        = ( isset( $attributes['url'] ) ) ? $attributes['url'] : false;
-	$label      = ( isset( $attributes['label'] ) ) ? $attributes['label'] : block_core_social_link_get_name( $service );
-	$class_name = isset( $attributes['className'] ) ? ' ' . $attributes['className'] : false;
+	$service         = ( isset( $attributes['service'] ) ) ? $attributes['service'] : 'Icon';
```
Comment (apeatling) [inline packages/block-library/src/social-link/index.php:19]: I think this and the others could be updated to use `array_key_exists( 'service', $attributes )`. That seems to be a cleaner way that some of the other core blocks now use.

File: packages/block-library/src/social-links/edit.js (hunk 1 above)
Comment (glendaviesnz) [reply packages/block-library/src/social-links/edit.js]: The switch to optional chaining has fixed this issue - @apeatling are you happy to clear this?

**Comment by glendaviesnz** [comment]
This all seems to work as advertised. 

Apologies if you already explored this, but the specificity of the individual link color settings is reasonably light, so rather than having to pass down context and set individual attributes for each embedded link is there some we we could instead add a generic style at the parent level? eg. adding a style of:
```css
.override-background.wp-block-social-links:not(.is-style-logos-only) .wp-social-link {
    background-color: red;
}
```
replaces the background on all the child icons.

**Comment by aaronrobertshaw** [comment]
> Apologies if you already explored this, but the specificity of the individual link color settings is reasonably light, so rather than having to pass down context and set individual attributes for each embedded link is there some way we could instead add a generic style at the parent level? 

That was my thinking initially as well, to be able to just apply the color choices to the social links block itself and allow the individual links to inherit that as required. Unfortunately, I couldn't find a way to achieve the desired results doing this.

Some of the issues I ran into in this area were:
* Maintaining consistency by allowing the normal use of color CSS classes e.g. `has-accent-background-color`
* Not being able to use those color classes on the container without them being applied but still inheritable (e.g. background colors).
* As the color palette could be defined by the theme, I couldn't create a new set of styles setting the correct color like the example you gave
* Custom colors in general using inline styles make the background colors on the parent a problem again.
* Still maintaining the individual social media service colors without blowing out the CSS footprint

I could be missing something so if you do have further suggestions, let me know and I can explore those. For now though, this does give us the control over colors we need, hopefully that is ok.

**Comment by glendaviesnz** [comment]
> I could be missing something so if you do have further suggestions, let me know and I can explore those. For now though, this does give us the control over colors we need, hopefully that is ok.

No, I don't think you are missing anything, it just seemed like there should be a simpler approach, but due to the factors you list I think the approach you have take is the only way to push the styles down to the children.

**Comment by glendaviesnz** [comment]
@apeatling do you have any suggestions on who is best to give us a signoff  on this approach?

**Comment by apeatling** [comment]
Maybe @nosolosw or @youknowriad might have thoughts on the approach?

**Comment by youknowriad** [comment]
Hi there! thanks for the PR.

In general, we're trying to move away from specific customization and attributes in favor of generic support flags that can be enabled across blocks, so my questions here are:

Can we use the `__experimentalColor` support flag for the background color support here instead? It does support text and backgrouund color and it seems the link color can be considered as the same thing as "icon color"? Can't we just provide an option to change the "label"?

**Comment by aaronrobertshaw** [comment]
Hi @youknowriad, thanks for taking a look at this! 

I'm still trying to get my head around how everything fits together. I did initially try to leverage `__experimentalColor` however I ran into problems applying the colors only to each child within the Social Links' `InnerBlocks`. 

Looking back at this PR, do you think an option would be to use `__experimentalColors` without wrapping the `InnerBlocks` with the generated color components. Then still make those color attributes available to the inner links via the block context?

**Comment by youknowriad** [comment]
I was thinking `__experimentalColors` could be applied to SocialLink and not SocialLinks? or is this not good enough in terms of UX?

**Comment by aaronrobertshaw** [comment]
The idea was, the better UX was for the user to only have to make a single set of color selections. That is the approach used by the navigation block.

**Comment by youknowriad** [comment]
I wonder if there's a way to make the settings on the parent component work as a "default value" for its children instead.

**Comment by nosolosw** [comment]
@youknowriad I don't have an answer for this at the moment but logged this example at https://github.com/WordPress/gutenberg/issues/22887 because I smell that these little glitches in the system may be addressed together.

**Comment by aaronrobertshaw** [comment]
@youknowriad and @nosolosw would you be happy for us to move forward with this current approach so we have the functionality in place? I'm happy to come back and revisit this when we have a solution to the larger questions raised.

**Comment by youknowriad** [comment]
Personally, I'd rather compromise on the UI and make this per icon for now until we figure out a better API for handling `parent -> child attributes UI`. What do you think here @mtias

**Comment by mtias** [comment]
It doesn't make sense per icon to me

**Comment by aaronrobertshaw** [comment]
@apeatling Would you like to weigh in and cast a vote on a preferred approach to this one?

Should we allow what is the best experience for a user, to be the tie breaker if needed?

**Comment by apeatling** [comment]
I agree that changing per icon is not the ideal user experience, and we're better off optimizing for the user.

**Comment by aaronrobertshaw** [comment]
An alternative approach to this is available in https://github.com/WordPress/gutenberg/pull/28084

I'm inclined to close this PR in favour of the new approach if others agree.

**Comment by aaronrobertshaw** [comment]
Closing this PR in favour of https://github.com/WordPress/gutenberg/pull/28084

**Pull request closed without merging**